spring-kafka/src/test/java/org/springframework/kafka/config/KafkaListenerEndpointRegistryTests.java

```diff
@@ -47,4 +60,123 @@ void unregister() {
+			Arguments.of(names, (Predicate<String>) id -> false, 0)
+		);
+	}
+
+	@ParameterizedTest(name = "getListenerContainersMatching with BiPredicate for {0}, expecting {2} matches")
+	@MethodSource("paramsForGetListenerContainersMatchingBiPredicate")
+	void getListenerContainersMatchingBiPredicate(List<String> names, BiPredicate<String, MessageListenerContainer> idAndContainerMatcher, int expectedCount) {
+		// Given
+		KafkaListenerEndpointRegistry registry = new KafkaListenerEndpointRegistry();
+		registerWithListenerIds(registry, names);
+		// When
+		Collection<MessageListenerContainer> listeners = registry.getListenerContainersMatching(idAndContainerMatcher);
+		// Then
+		assertThat(listeners).hasSize(expectedCount);
+	}
+
+	/**
+	 * Provides parameters for the getListenerContainersMatchingBiPredicate test.
+	 * Each set of parameters includes a list of names, a bi-predicate, and the expected count of matching containers.
+	 */
+	private static Stream<Arguments> paramsForGetListenerContainersMatchingBiPredicate() {
+		List<String> names = List.of("foo", "bar", "baz");
+		return Stream.of(
+			// Case : Filters for names starting with "b" and containers that are "running"
+			Arguments.of(names,
+				(BiPredicate<String, MessageListenerContainer>) (id, container) -> id.startsWith("b") && container.isRunning(), 2),
+			// Case : Filters for names starting with "f" and containers that are "running"
+			Arguments.of(names,
+				(BiPredicate<String, MessageListenerContainer>) (id, container) -> id.startsWith("f") && container.isRunning(), 1),
+			// Case : Filters in an empty list of names
+			Arguments.of(new ArrayList<>(),
+				(BiPredicate<String, MessageListenerContainer>) (id, container) -> id.startsWith("b") && container.isRunning(), 0),
+			// Case : Filters where all containers are considered "running"
+			Arguments.of(names,
+				(BiPredicate<String, MessageListenerContainer>) (id, container) -> container.isRunning(), 3),
+			// Case : Filters where no containers are considered "running"
+			Arguments.of(names,
+				(BiPredicate<String, MessageListenerContainer>) (id, container) -> !container.isRunning(), 0)
+		);
+	}
+
+	private static void registerWithListenerIds(KafkaListenerEndpointRegistry registry, List<String> names) {
+		names.forEach(name -> registerListenerWithId(registry, name));
+	}
+
+	private static void registerListenerWithId(KafkaListenerEndpointRegistry registry, String id) {
+		KafkaListenerEndpoint endpoint = mock(KafkaListenerEndpoint.class);
+		@SuppressWarnings("unchecked")
+		KafkaListenerContainerFactory<MessageListenerContainer> factory = mock(KafkaListenerContainerFactory.class);
+		given(endpoint.getId()).willReturn(id);
+		MessageListenerContainer container = mock(MessageListenerContainer.class);
+		given(container.isRunning()).willReturn(true);
+		given(factory.createListenerContainer(endpoint)).willReturn(container);
+		registry.registerListenerContainer(endpoint, factory);
+	}
```

Comment:
Every member (including the last method of the last class) in the class must be surrounded with blank lines for much better code readability.